### ISSUE TITLE
[Java] UseConcurrentHashMap false positive for locally used map.

### ISSUE THREAD

**Issue by mnlipp** [pmd/pmd#1022]
**Affects PMD Version:** 6.2.0

**Rule:** UseConcurrentHashMap

**Description:** Using a `HashMap` with purely local scope triggers this warning, although it is not accessible to another thread. This is especially annoying because the [documentation](https://pmd.github.io/pmd-6.2.0/pmd_rules_java_multithreading.html#useconcurrenthashmap) shows this as fine:
```
Map map1 = new HashMap();           // fine for single-threaded access
```

**Code Sample demonstrating the issue:**

```
import java.util.HashMap;
import java.util.Map;

public class Test {

	public Test() {
		Map<Integer,Integer> map = new HashMap<>();
		for (int i = 0; i < 10; i++) {
			map.put(i, i);
		}
	}

}

```

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]* eclipse-pmd 1.11

**Comment by jsotuyod** [comment]
@mnlipp thanks for your report. Unfortunately, this rule doesn't do what you expect it to do.

Without complex multifile analysis, it's impossible for the rule to know if a given object instance is shared across multiple threads, so it will simply flag all maps that are not `ConcurrentHashMap`, and the rule is basically intended to be a reminder to use it everywhere (not necessarily the best decision, so use the rule at your own discretion).

I believe we should do two changes:
1. Update the documentation to better reflect this
2. Update the rule to at least not flag local instances, as they for sure are not shared among threads. Fields and parameters would still be flagged unconditionally 'though.
